vkubelet/pod.go

```diff
@@ -94,25 +98,94 @@ func (s *Server) deletePod(ctx context.Context, namespace, name string) error {
 		return s.forceDeletePodResource(ctx, namespace, name)
 	}
 
-	ctx, span := trace.StartSpan(ctx, "deletePod")
+	ctx, span := trace.StartSpan(ctx, "deletePodFromProvider")
 	defer span.End()
 	ctx = addPodAttributes(ctx, span, pod)
 
-	var delErr error
-	if delErr = s.provider.DeletePod(ctx, pod); delErr != nil && errors.IsNotFound(delErr) {
+	if delErr := s.provider.DeletePod(ctx, pod); delErr != nil && errors.IsNotFound(delErr) {
 		span.SetStatus(ocstatus.FromError(delErr))
 		return delErr
 	}
 
 	log.G(ctx).Debug("Deleted pod from provider")
 
-	if !errors.IsNotFound(delErr) {
-		if err := s.forceDeletePodResource(ctx, namespace, name); err != nil {
-			span.SetStatus(ocstatus.FromError(err))
-			return err
-		}
-		log.G(ctx).Info("Deleted pod from Kubernetes")
+	return nil
+}
+
+// terminatePod terminates the pod in the provider. It assumes the API
+// server pod has been marked for deletion
+func (s *Server) terminatePod(ctx context.Context, pod *corev1.Pod) error {
+	ctx, span := trace.StartSpan(ctx, "terminatePod")
+	defer span.End()
+	ctx = addPodAttributes(ctx, span, pod)
+	ctx = span.WithField(ctx, "softDeletes", s.softDeletes)
+
+	if !s.softDeletes {
+		return s.deletePod(ctx, pod.Namespace, pod.Name)
+	}
+
+	// Check if the pod has already been soft-deleted
+	if _, ok := pod.ObjectMeta.Annotations[softDeleteKey]; ok {
+		return nil
+	}
+
+	err := s.deletePodFromProvider(ctx, pod.Namespace, pod.Name)
+	if err != nil && !errors.IsNotFound(err) {
+		log.G(ctx).WithError(err).Warn("Failed to delete pod from provider")
+		span.SetStatus(ocstatus.FromError(err))
+		return err
+	}
+	// If the pod deletion was successful, or the provider did not know about the pod
+	// then go ahead and tombstone the pod, otherwise we allow the pod to persist in
+	// API server, and if it gets another update, we will retry deleting it
+	// from the provider.
+	//
+	// Reasoning: A provider may get into an inconsistent state, and a restart might
+	// be necessary go get it back into a good state. We would rather allow for graceful
+	// termination if possible
```

Comment:
So assuming the VK and/or provider restarted, where in the code exactly will the reconciliation happen?